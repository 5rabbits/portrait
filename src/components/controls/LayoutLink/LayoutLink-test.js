import { React, toJSON, mount } from 'test/helper'
import LayoutLink from 'controls/LayoutLink'

describe('LayoutLink', () => {
  it('should render with minimum props', () => {
    const component = mount(
      <LayoutLink>test</LayoutLink>,
    )

    expect(toJSON(component)).toMatchSnapshot()
  })

  describe('handleClick(event)', () => {
    it('should prevent default', () => {
      const component = mount(
        <LayoutLink>test</LayoutLink>,
      )
      const instance = component.instance()
      const event = { preventDefault: jest.fn() }

      instance.handleClick(event)

      expect(event.preventDefault).toHaveBeenCalledTimes(1)
    })

    describe('if props.onClick is defined', () => {
      it('should invoke props.onClick passing the event', () => {
        const handleClick = jest.fn()
        const component = mount(
          <LayoutLink onClick={handleClick}>test</LayoutLink>,
        )
        const instance = component.instance()
        const event = { preventDefault: jest.fn() }

        instance.handleClick(event)

        expect(handleClick).toHaveBeenCalledTimes(1)
        expect(handleClick).toHaveBeenCalledWith(event)
      })
    })
  })

  describe('if props.href is "#"', () => {
    it('should assign this.handleClick to the underlying anchor', () => {
      const component = mount(
        <LayoutLink href="#">test</LayoutLink>,
      )
      const instance = component.instance()

      expect(component.find('a').prop('onClick')).toBe(instance.handleClick)
    })
  })

  describe('if props.href is not "#"', () => {
    it('should assign props.onClick to the underlying anchor', () => {
      const handleClick = jest.fn()
      const component = mount(
        <LayoutLink
          href="/something"
          onClick={handleClick}
          >
          test
        </LayoutLink>,
      )

      expect(component.find('a').prop('onClick')).toBe(handleClick)
    })
  })
})
