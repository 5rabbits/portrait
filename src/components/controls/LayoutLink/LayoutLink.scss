@import '~styles/core';

$transition-duration: 200ms;

.LayoutLink {
  border-radius: 20px;
  padding: 4px;
  transition: opacity $transition-duration ease-out,
    padding $transition-duration ease-out,
    background $transition-duration ease-out;
  font-size: $font-size-footnote;
  line-height: $line-height-footnote;
  white-space: nowrap;

  &.LayoutLink--active,
  &:hover {
    background: rgba($primary-color, 0.1);
    padding: 4px 10px;
    text-decoration: none;
  }
}
