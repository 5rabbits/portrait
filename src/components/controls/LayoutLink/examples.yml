- description: Default state.
  code: >
    <LayoutLink>Toggle panel</LayoutLink>

- description: Active state.
  code: >
    <LayoutLink active>Toggle panel</LayoutLink>

- description: Disabled.
  code: >
    <LayoutLink disabled>Disabled link</LayoutLink>
