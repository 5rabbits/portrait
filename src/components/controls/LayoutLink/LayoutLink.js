import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'
import './LayoutLink.scss'

/**
 * Link used to display layout changes on the main section containers.
 */
export default class LayoutLink extends PureComponent {
  static propTypes = {
    /**
     * Whether the link should be displayed with active/hover styles.
     */
    active: PropTypes.bool,

    /**
     * The link contents to be displayed.
     */
    children: PropTypes.node.isRequired,

    /**
     * CSS class assigned to the container.
     */
    className: PropTypes.string,

    /**
     * Whether the link is disabled or not.
     */
    disabled: PropTypes.bool,

    /**
     * Optional link href.
     */
    href: PropTypes.string,

    /**
     * Callback invoked when the user clicks the link.
     */
    onClick: PropTypes.func,
  }

  static defaultProps = {
    active: false,
    className: null,
    disabled: false,
    href: '#',
    onClick: null,
  }

  handleClick = event => {
    event.preventDefault()

    if (this.props.disabled) {
      return
    }

    if (this.props.onClick) {
      this.props.onClick(event)
    }
  }

  render() {
    const { active, children, className, disabled, href, onClick, ...other } = this.props

    return (
      <a
        className={cx('LayoutLink', className, {
          'LayoutLink--active': active,
          'LayoutLink--disabled': disabled,
        })}
        href={href}
        onClick={(href === '#' || disabled) ? this.handleClick : onClick}
        {...other}
        >
        {children}
      </a>
    )
  }
}
