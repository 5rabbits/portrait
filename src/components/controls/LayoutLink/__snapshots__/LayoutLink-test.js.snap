// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoutLink should render with minimum props 1`] = `
<LayoutLink
  active={false}
  className={null}
  disabled={false}
  href="#"
  onClick={null}
>
  <a
    className="LayoutLink"
    href="#"
    onClick={[Function]}
  >
    test
  </a>
</LayoutLink>
`;
